subsys/net/lib/ocpp/core.c

```diff
@@ -0,0 +1,504 @@
+		enum ocpp_key_type ktype;
+
+		ktype = ocpp_get_keyval_type(key);
+		if (ktype < KEY_TYPE_STR) {
+			kval.ival = atoi(sval);
+		} else {
+			kval.str = sval;
+		}
+		ret = ocpp_update_cfg_val(key, &kval);
+
+		if (ret) {
+			res = "Rejected";
+		}
+	} else {
+		res = "NotSupported";
+	}
+
+	if (!ret) {
+		switch (key) {
+		case CFG_MTR_VAL_SAMPLE_INTERVAL:
+			if (atomic_get(&ctx->mtr_timer_ref_cnt) <= 0) {
+				break;
+			}
+
+			k_timer_start(&ctx->mtr_timer,
+				      K_SECONDS(kval.ival),
+				      K_SECONDS(kval.ival));
+			break;
+
+		default:
+			break;
+		}
+	}
+
+	fn = ctx->cfn[PDU_CHANGE_CONFIGURATION];
+	fn(buf, sizeof(ctx->pdu_buf), res, uid);
+
+	rmsg.ctx = ctx;
+	rmsg.msg = buf;
+	rmsg.msg_len = strlen(buf) + 1;
+
+	ret = ocpp_send_to_server(&rmsg, K_SECONDS(OCPP_PDU_TIMEOUT));
+
+	return ret;
+}
+
+int ocpp_authorize(ocpp_session_handle_t hndl, char *idtag,
+		   enum ocpp_auth_status *status,
+		   uint32_t timeout_ms)
+{
+	int ret;
+	struct ocpp_session *sh = (struct ocpp_session *)hndl;
+	struct ocpp_info *ctx;
+	struct ocpp_upstream_info *ui;
+	char buf[256];
+	union ocpp_keyval *val;
+	struct ocpp_wamp_rpc_msg rmsg = {0};
+	ocpp_msg_fp_t fn;
+
+	if (!idtag || !status ||
+	    !ocpp_session_is_valid(sh)) {
+		return -EINVAL;
+	}
+
+	ctx = sh->ctx;
+	if (ctx->state < CP_STATE_READY) {
+		return -EAGAIN;
+	}
+
+	if (ctx->is_cs_offline) {
+		val = ocpp_get_key_val(CFG_LOCAL_AUTH_OFFLINE);
+		if (val && !val->ival) {
+			return -EAGAIN;
+		}
+	}
+
+	val = ocpp_get_key_val(CFG_AUTH_CACHE_ENABLED);
+	if (val && val->ival) {
+
+	}
+
+	strncpy(sh->idtag, idtag, sizeof(sh->idtag));
+	ui = ctx->ui;
+	fn = ctx->cfn[PDU_AUTHORIZE];
+	sh->uid = fn(buf, sizeof(buf), sh);
+
+	rmsg.ctx = ctx;
+	rmsg.msg = buf;
+	rmsg.msg_len = strlen(buf) + 1;
+	rmsg.sndlock = &ui->ws_sndlock;
+	rmsg.rspsig = &ui->ws_rspsig;
+	ret = ocpp_send_to_server(&rmsg, K_MSEC(timeout_ms));
+	if (ret < 0) {
+		return ret;
+	}
+
+	*status = sh->resp_status;
+	if (sh->resp_status == OCPP_AUTH_ACCEPTED) {
+		sh->is_active = true;
+	}
+
+	return 0;
+}
+
+int ocpp_heartbeat(ocpp_session_handle_t hndl)
+{
+	int ret;
+	struct ocpp_session *sh = (struct ocpp_session *)hndl;
+	struct ocpp_info *ctx = sh->ctx;
+	struct ocpp_upstream_info *ui = ctx->ui;
+	ocpp_msg_fp_t fn;
+	struct ocpp_wamp_rpc_msg rmsg = {0};
+	char *buf = ctx->pdu_buf;
+
+	fn = ctx->cfn[PDU_HEARTBEAT];
+	sh->uid = fn(buf, sizeof(ctx->pdu_buf), sh);
+
+	rmsg.ctx = ctx;
+	rmsg.msg = buf;
+	rmsg.msg_len = strlen(buf) + 1;
+	rmsg.sndlock = &ui->ws_sndlock;
+	rmsg.rspsig = &ui->ws_rspsig;
+
+	ret = ocpp_send_to_server(&rmsg, K_SECONDS(OCPP_PDU_TIMEOUT));
+	if (!ret && sh->resp_status != BOOT_ACCEPTED) {
+		ret = -EAGAIN;
+	}
+
+	return ret;
+}
+
+int ocpp_start_transaction(ocpp_session_handle_t hndl,
+			   int Wh,
+			   uint8_t conn_id,
+			   uint32_t timeout_ms)
+{
+	struct ocpp_session *sh = (struct ocpp_session *)hndl;
+	struct ocpp_info *ctx;
+	struct ocpp_upstream_info *ui;
+	char buf[350];
```

Comment:
I am a bit worried about these hard coded temp buffers from stack. You have set the stack size quite large which helps here. Why do we need such a large buffer, and could we have a `define` for the size instead of hard coding the value.